store/cachekv/store.go

```diff
@@ -114,22 +114,34 @@ func (store *Store) Write() {
 	// TODO: Consider allowing usage of Batch, which would allow the write to
 	// at least happen atomically.
 	for _, key := range keys {
-		cacheValue := store.cache[key]
-
-		switch {
-		case store.isDeleted(key):
+		if store.isDeleted(key) {
+			// We use []byte(key) instead of conv.UnsafeStrToBytes because we cannot
+			// be sure if the underlying store might do a save with the byteslice or
+			// not. Once we get confirmation that .Delete is guaranteed not to
+			// save the byteslice, then we can assume only a read-only copy is sufficient.
 			store.parent.Delete([]byte(key))
-		case cacheValue.value == nil:
-			// Skip, it already doesn't exist in parent.
-		default:
+			continue
+		}
+
+		cacheValue := store.cache[key]
+		if cacheValue.value != nil {
+			// It already exists in the parent, hence delete it.
 			store.parent.Set([]byte(key), cacheValue.value)
 		}
 	}
 
-	// Clear the cache
-	store.cache = make(map[string]*cValue)
-	store.deleted = make(map[string]struct{})
-	store.unsortedCache = make(map[string]struct{})
+	// Clear the cache using the map clearing idiom
+	// and not allocating fresh objects.
+	// Please see https://bencher.orijtech.com/perfclinic/mapclearing/
+	for key := range store.cache {
+		delete(store.cache, key)
+	}
+	for key := range store.deleted {
+		delete(store.deleted, key)
+	}
+	for key := range store.unsortedCache {
+		delete(store.unsortedCache, key)
+	}
```

Comment:
I think this is incorrect in the general case. I've had to actually undo this in the CacheKV store myself in the Osmosis branch in a couple places.

This preserves the old capacity of the map, which makes iterating over map entries / sorting / etc. take much longer.

So in the Osmosis case, or a chain upgrading case, if you have one very large blcok, then many short blocks, every short block will take similar store time for iteration as the long block does.

What we did in the Osmosis branch was only use a map clearing idiom if the map is sufficiently small, otherwise we just clear it.

(This made a difference between N^2 behavior and O(N) behavior in our codebase, due to iterator things. Potentially less crucial here, but I'd still caution against this as a premature optimization)

Reply:
According to https://go-review.googlesource.com/c/go/+/110055/, the struct reinitialized and memory is cleared. So the optimization is that we save on heap memory allocation. However, if one TX will consume looooot of memory then we may have a problem (eg: huge iterator).

Reply:
@odeke-em can you help answer ^^